Explanation for SetLabel/Annotation prefixes

markmandel opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://github.com/googleforgames/agones/blob/main/CONTRIBUTING.md and developer guide https://github.com/googleforgames/agones/blob/main/build/README.md
2. Please label this pull request according to what type of issue you are addressing.
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/googleforgames/agones/blob/main/build/README.md#testing-and-building
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup

/kind documentation

> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:

Better explantations for the reasons why both `SetLabel(...)` and `SetAnnotation(...)` on the game server SDK are prefixed with values when being set on the GameServer.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Closes #<issue number>`, or `Closes (paste link of issue)`.
-->
Closes #2053

**Special notes for your reviewer**:

N/A

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
-record that is stored in Kubernetes. To maintain isolation, the `key` value is automatically prefixed with "agones.dev/sdk-".
+record that is stored in Kubernetes. 
+
+To maintain isolation, the `key` value is automatically prefixed with value **"agones.dev/sdk-"**. This is done for two 
```
nit: "with the value"

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
+To maintain isolation, the `key` value is automatically prefixed with value **"agones.dev/sdk-"**. This is done for two 
+main reasons:
+*  The prefix allows the developer to always know if they are accessing or reading a value that could have come, or 
+   may be changed by the client SDK. Much like `private` vs `public` scope in a programming language - The Agones SDK only 
```
nit: change - to , and The to the

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
+main reasons:
+*  The prefix allows the developer to always know if they are accessing or reading a value that could have come, or 
+   may be changed by the client SDK. Much like `private` vs `public` scope in a programming language - The Agones SDK only 
+   gives you access to part of the set of labels and annotations that exist on a GameServer.
```
nit: gives you access to write to part of the set of labels...

domgreen replied:
agree with this clearly point out that you can edit only the labels that have the prefix.

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
+   gives you access to part of the set of labels and annotations that exist on a GameServer.
+*  The prefix allows for a smaller attack surface if the GameServer container gets compromised. Since the 
+   game container is generally externally exposed, and the Agones project doesn't control the binary that is 
+   run within it - limiting exposure if something goes horribly wrong, and it becomes compromised is worth the extra 
```
nit: change - to ,

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
+   gives you access to part of the set of labels and annotations that exist on a GameServer.
+*  The prefix allows for a smaller attack surface if the GameServer container gets compromised. Since the 
+   game container is generally externally exposed, and the Agones project doesn't control the binary that is 
+   run within it - limiting exposure if something goes horribly wrong, and it becomes compromised is worth the extra 
```
nit: remove , after wrong

markmandel replied:
grammar check on IntelliJ tells me to comma after wrong 😄

roberthbailey replied:
Maybe that's regex matching because it is following by and? The way I read it "something goes horribly wrong and it becomes compromised" is a single phrase, not two phrases that need to be separated by a comma. But maybe my pattern matching is off....

markmandel replied:
removed the "horribly wrong" bit - also reads cleaner and less emotionally charged language 😄

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
 
 #### SetAnnotation(key, value)
 
 This will set a [Annotation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) value on the backing
```
nit: "an Annotation"

roberthbailey commented on site/content/en/docs/Guides/Client SDKs/_index.md:
```diff
@@ -183,21 +183,38 @@ please [file an issue](https://github.com/googleforgames/agones/issues).
+`GameServer` record that is stored in Kubernetes. 
 
-This can be useful if you want information from your running game server process to be observable through the Kubernetes API.
+To maintain isolation, the `key` value is automatically prefixed with **"agones.dev/sdk-"**. Since Agones also uses 
```
To maintain isolation, the `key` value is automatically prefixed with **"agones.dev/sdk-"** for the same reasons as in [SetLabel(...)](#setlabelkey-value) above. The isolation is also important as Agones uses annotations on the `GameServer` as part of its internal processing.

roberthbailey replied:
For "internal processing" do we want to say anything more specific, like game state lifecycle?

markmandel replied:
I deliberately left it ambiguous: (a) so people don't rely on things we do and (b) in case it ever changes, we don't have to ever go back and alter the documentation.

roberthbailey replied:
That's ok. Ending the sentence with "as part of its internal processing." didn't feel super satisfying to me but I couldn't figure out anything else to add that would make sense either.

roberthbailey reviewed (commented):
Lots of grammar nits (sorry). Overall a definite improvement!

domgreen reviewed (commented):
Wow, your quick :smile: 

I like the additions it clearly points out why the prefix is present :+1: 

My only addition would be something (possibly an example) that states that if you wish to set a label (possibly via allocation) you should add the prefix to any labels/annotations that you wish the GS to modify eg:

```
Labels and annotations can be added to a GameServer from outside the process such as via Allocation or kubectl. If you wish the GameServer to be able to edit these ensure that they are prefixed with `agones.dev/sdk-` so that the GameServer has access to editing this data.

Additionally be aware that majority of the SDKs will automatically add `agones.dev/sdk-` to any labels that are modified for example in the Unreal SDK:

AgonesSDK->SetLabel("foo", "bar", {})  // resulting label agones.dev/sdk-foo: bar
```

markmandel commented:
Excellent comments! Updated.

@domgreen I added a section to the FAQ that I think covers this scenario. PTAL.

Long term, I do want to add some kind of "Allocation Guide" that covers different scenarios, especially with the work coming from #1239.

roberthbailey commented:
@domgreen - once you take another look we'll get this merged.

domgreen commented on site/content/en/docs/FAQ/_index.md:
```diff
@@ -79,6 +79,11 @@ through [SDK.WatchGameServer()]({{< ref "/docs/Guides/Client SDKs/_index.md#watc
 Combining these two features allows you to pass information such as map data, gameplay metadata and more to a game
 server binary at Allocation time, through Agones functionality.
 
+Do note, that if you wish to have either the labels or annotations that are set via `GameServerAllocation` to be 
+editable by the game server via the Agones SDK, the label key will need to be prefixed with `agones.dev/sdk-`. See
```
this doesnt read right in my head at least...

```suggestion
Do note, that if you wish to have either the labels or annotations set by `GameServerAllocation` to ensure these are editable by the game server from the Agones SDK, the label key will need to be prefixed with `agones.dev/sdk-`. See
```

domgreen replied:
Additional comment ... do we refer to gs as `gameservers` `game server` `Game Server`? I tend to refer to it as DGS dedicated game server :joy:

markmandel replied:
I usually use `GameServer` when referring to the CRD, and game server when talking about the binary.

markmandel replied:
Tweaked the wording a bit - I think I solved the awkward phrasing. Thanks!

domgreen reviewed (approved):
I'm more than happy with the addition small suggestion about readability on the FAQ (multiple via read strange for me) but only a suggestion.

I really like being able to see the built out version of the site :+1:
